<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Azimuth annotation • Azimuth</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Azimuth annotation">
<meta property="og:description" content="Azimuth">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Azimuth</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.4.5.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/run_azimuth_tutorial.html">Azimuth annotation</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/satijalab/azimuth/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Azimuth annotation</h1>
            
            <h4 data-toc-skip class="date">Compiled: May 10, 2022</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/satijalab/azimuth/blob/HEAD/vignettes/run_azimuth_tutorial.Rmd" class="external-link"><code>vignettes/run_azimuth_tutorial.Rmd</code></a></small>
      <div class="hidden name"><code>run_azimuth_tutorial.Rmd</code></div>

    </div>

    
    
<hr>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>This vignette demonstrates how to map scRNA-seq data to Azimuth reference datasets in R.</p>
<p>Reference-based mapping provides an attractive alternative to unsupervised analysis. When well-curated and annotated references are available, reference-mapping can rapdidly, robustly, and sensitively annotate and interpret query datasets. As part of the <a href="portal.hubmapconsortium.org">Human Biomolecular Atlas Project</a>, we have built integrated references for multiple human tissues, available at <a href="azimuth.hubmapconsortium.org">azimuth.hubmapconsortium.org</a>. Azimuth is a web-tool that maps user-uploaded datasets - starting from an unnormalized expression counts matrix.</p>
<p>We have recently released updated software that enables the Azimuth workflow to be replicated directly in R, without the need to upload data into the web application. Reference datasets are automatically downloaded as part of out previously released <a href="https://github.com/satijalab/seurat-data" class="external-link">SeuratData</a> framework. Datasets can then be passed through Azimuth using a single command.</p>
<p>As with the web application, Azimuth is compatible with a wide range of inputs, including Seurat objects, 10x HDF5 files, and Scanpy/h5ad files. Once Azimuth is run, a Seurat object is returned which contains</p>
<ul>
<li>Cell annotations (at multiple levels of resolution)</li>
<li>Prediction scores (i.e. confidence scores) for each annotation</li>
<li>Projection onto the reference-derived 2-dimensional UMAP visualization</li>
</ul>
<p>In this vignette, we demonstrate the use of a function <code><a href="../reference/RunAzimuth.html">RunAzimuth()</a></code> which facilitates annotation of single cell datasets.</p>
<details><summary><strong>View data download code</strong>
</summary><p>To download all the required files, you can run the following lines in a shell:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://figshare.com/ndownloader/files/27686835 <span class="at">-O</span> human_cd34_bone_marrow.h5ad</span></code></pre></div>
</details>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>We first install and load Seurat, Azimuth, and Seurat-Data</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/remote-reexports.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"satijalab/seurat-data"</span><span class="op">)</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/remote-reexports.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"satijalab/azimuth"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"Seurat"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/satijalab/azimuth" class="external-link">Azimuth</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://www.satijalab.org/seurat" class="external-link">SeuratData</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></code></pre></div>
<div class="section level3">
<h3 id="map-pbmc-scrna-seq-datasets-from-7-different-technologies">Map PBMC scRNA-seq datasets from 7 different technologies<a class="anchor" aria-label="anchor" href="#map-pbmc-scrna-seq-datasets-from-7-different-technologies"></a>
</h3>
<p>We first consider a benchmark dataset from <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7289686/" class="external-link">Systematic comparison of single-cell and single-nucleus RNA-sequencing methods, Nat. Biotech 2020</a>, which profiled a total of 31,021 human peripheral blood mononuclear cells (PBMCs) using seven different experimental technologies. The raw data is avaialble for download <a href="https://singlecell.broadinstitute.org/single_cell/study/SCP424/single-cell-comparison-pbmc-data" class="external-link">here</a>, but the dataset is also available for easy loading via SeuratData. We load the data in, and map it to our <a href="https://azimuth.hubmapconsortium.org/references/#Human%20-%20PBMC">PBMC reference</a>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Install the PBMC systematic comparative analyis (pmbcsca) dataset</span>
<span class="fu"><a href="https://satijalab.org/seurat/reference/InstallData.html" class="external-link">InstallData</a></span><span class="op">(</span><span class="st">"pbmcsca"</span><span class="op">)</span>

<span class="co"># returns a Seurat object named pbmcsca</span>
<span class="fu"><a href="https://satijalab.org/seurat/reference/LoadData.html" class="external-link">LoadData</a></span><span class="op">(</span><span class="st">"pbmcsca"</span><span class="op">)</span>

<span class="co"># The RunAzimuth function can take a Seurat object as input</span>
<span class="va">pbmcsca</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RunAzimuth.html">RunAzimuth</a></span><span class="op">(</span><span class="va">pbmcsca</span>, reference <span class="op">=</span> <span class="st">"pbmcref"</span><span class="op">)</span></code></pre></div>
<p>We can now visualize the outputs of Azimuth. Note that all cells are labeled with high-resolution annotations, and are projected into a harmonized space despite being analyzed from a wide variety of technologies.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">pbmcsca</span>, group.by <span class="op">=</span> <span class="st">"predicted.celltype.l2"</span>, label <span class="op">=</span> <span class="cn">TRUE</span>, label.size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">pbmcsca</span>, group.by <span class="op">=</span> <span class="st">"Method"</span><span class="op">)</span>
<span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></code></pre></div>
<p><img src="run_azimuth_tutorial_files/figure-html/pbmc_dimplot-1.png" width="1152"></p>
<p>We can visualize the expression of canonical marker genes to examine the accuracy of predictions. Note that Azimuth normalizes data (internally) before mapping, but does not return the results, so we normalize the data here before visualization.</p>
<p>Here, we specifically visualize:</p>
<ul>
<li>The expression of CCR7 on CD4 and CD8 Naive T cells</li>
<li>The expression of FCGR3A on CD16+ monocytes, CD56dim NK cells, and cytotoxic CD8 T cells</li>
<li>The expression of AXL on rare populations of AXL+SIGLEC6+ dendritic cells (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5775029/" class="external-link">ASDC</a>)</li>
<li>Prediction scores for the annotation CD4+ regulatory T cells (Treg)</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmcsca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">pbmcsca</span><span class="op">)</span>
<span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">pbmcsca</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"predicted.celltype.l2"</span>

<span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">pbmcsca</span>, features <span class="op">=</span> <span class="st">"CCR7"</span><span class="op">)</span>
<span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">pbmcsca</span>, features <span class="op">=</span> <span class="st">"FCGR3A"</span><span class="op">)</span>
<span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html" class="external-link">VlnPlot</a></span><span class="op">(</span><span class="va">pbmcsca</span>, features <span class="op">=</span> <span class="st">"AXL"</span>, group.by <span class="op">=</span> <span class="st">"predicted.celltype.l2"</span>, idents <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ASDC"</span>,
    <span class="st">"pDC"</span>, <span class="st">"cDC1"</span>, <span class="st">"cDC2"</span><span class="op">)</span><span class="op">)</span>
<span class="va">p4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">pbmcsca</span>, features <span class="op">=</span> <span class="st">"predictionscorecelltypel2_Treg"</span><span class="op">)</span>

<span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span> <span class="op">+</span> <span class="va">p3</span> <span class="op">+</span> <span class="va">p4</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></code></pre></div>
<p><img src="run_azimuth_tutorial_files/figure-html/pbmc_res1-1.png" width="1152"></p>
</div>
<div class="section level3">
<h3 id="see-available-references">See available references<a class="anchor" aria-label="anchor" href="#see-available-references"></a>
</h3>
<p>You can search all available datasets in SeuratData (focusing on Azimuth references)</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">available_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/AvailableData.html" class="external-link">AvailableData</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">available_data</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"Azimuth"</span>, <span class="va">available_data</span><span class="op">[</span>, <span class="fl">3</span><span class="op">]</span><span class="op">)</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">##                                  Dataset Version                        Summary</span>
<span class="co">## bonemarrowref.SeuratData   bonemarrowref   1.0.0  Azimuth Reference: bonemarrow</span>
<span class="co">## fetusref.SeuratData             fetusref   1.0.0       Azimuth Reference: fetus</span>
<span class="co">## humancortexref.SeuratData humancortexref   1.0.0 Azimuth Reference: humancortex</span>
<span class="co">## kidneyref.SeuratData           kidneyref   1.0.1      Azimuth Reference: kidney</span>
<span class="co">## lungref.SeuratData               lungref   2.0.0        Azimuth Reference: lung</span>
<span class="co">## mousecortexref.SeuratData mousecortexref   1.0.0 Azimuth Reference: mousecortex</span>
<span class="co">## pancreasref.SeuratData       pancreasref   1.0.0    Azimuth Reference: pancreas</span>
<span class="co">## pbmcref.SeuratData               pbmcref   1.0.0        Azimuth Reference: pbmc</span></code></pre>
</div>
<div class="section level3">
<h3 id="map-cd34-cells-from-human-bone-marrow-stored-as-an-h5ad-file">Map CD34+ cells from Human Bone Marrow (stored as an h5ad file)<a class="anchor" aria-label="anchor" href="#map-cd34-cells-from-human-bone-marrow-stored-as-an-h5ad-file"></a>
</h3>
<p>As a final example, we map data from CD34+ human bone marrow cells from the manuscript <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7549125/" class="external-link">Characterization of cell fate probabilities in single-cell data with Palantir</a>. We map to our <a href="https://azimuth.hubmapconsortium.org/references/#Human%20-%20Bone%20Marrow">Azimuth Human Bone Marrow reference</a>, which includes both progenitor and differentiated cells.</p>
<p>Azimuth can also take the path to an h5ad object as input. In this case, Azimuth extracts the unnormalized counts from the object, and proceeds with mapping.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RunAzimuth.html">RunAzimuth</a></span><span class="op">(</span>query <span class="op">=</span> <span class="st">"human_cd34_bone_marrow.h5ad"</span>, reference <span class="op">=</span> <span class="st">"bonemarrowref"</span><span class="op">)</span></code></pre></div>
<p>As expected, query cells map to CD34+ celltypes which represent a subset of celltypes present in the reference. Rare cells that map to differentiated cell populations (i.e. CD4 memory), map with low prediction scores. Reference-mapping also harmonizes two separate runs.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">bm</span>, group.by <span class="op">=</span> <span class="st">"predicted.celltype.l2"</span>, label <span class="op">=</span> <span class="cn">TRUE</span>, label.size <span class="op">=</span> <span class="fl">3</span>, repel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">bm</span>, group.by <span class="op">=</span> <span class="st">"orig.ident"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">bm</span><span class="op">$</span><span class="va">predicted.celltype.l2</span><span class="op">)</span>, decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## </span>
<span class="co">##           LMPP            GMP            HSC    Early Eryth            CLP </span>
<span class="co">##           1138           1065           1028            849            470 </span>
<span class="co">##          pre B     Late Eryth        pre-pDC            EMP transitional B </span>
<span class="co">##            422            214            152            141             82 </span>
<span class="co">##            pDC        Prog Mk          pro B         BaEoMa        pre-mDC </span>
<span class="co">##             68             60             38             18             14 </span>
<span class="co">##     CD4 Memory           cDC2           ASDC           MAIT        Stromal </span>
<span class="co">##              8              6              5              1              1</span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html" class="external-link">VlnPlot</a></span><span class="op">(</span><span class="va">bm</span>, features <span class="op">=</span> <span class="st">"predicted.celltype.l2.score"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span>
<span class="op">(</span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span><span class="op">)</span><span class="op">/</span><span class="va">p3</span></code></pre></div>
<p><img src="run_azimuth_tutorial_files/figure-html/bm_plots-1.png" width="960"></p>
<p>Lastly, we can visualize markers associated with lineage differentiation to verify our annotations including AVP (HSC), KLF1 (erythroid), MPO (myeloid), and VWF (platelet).</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># normalize before visualization</span>
<span class="va">bm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">bm</span><span class="op">)</span>
<span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">bm</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"AVP"</span>, <span class="st">"KLF1"</span>, <span class="st">"MPO"</span>, <span class="st">"VWF"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="run_azimuth_tutorial_files/figure-html/bm_feature_plots-1.png" width="1152"></p>
<details><summary><strong>Session Info</strong>
</summary><div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## R version 4.1.3 (2022-03-10)</span>
<span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">## Running under: Ubuntu 20.04.4 LTS</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0</span>
<span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span>
<span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] stats     graphics  grDevices datasets  utils     methods   base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">##  [1] patchwork_1.1.1                pbmcsca.SeuratData_3.0.0      </span>
<span class="co">##  [3] pbmcref.SeuratData_1.0.0       pbmc3k.SeuratData_3.1.4       </span>
<span class="co">##  [5] lungref.SeuratData_1.0.0       kidneyref.SeuratData_1.0.0    </span>
<span class="co">##  [7] bonemarrowref.SeuratData_1.0.0 SeuratData_0.2.2              </span>
<span class="co">##  [9] Azimuth_0.4.5                  shinyBS_0.61.1                </span>
<span class="co">## [11] sp_1.4-7                       SeuratObject_4.1.0            </span>
<span class="co">## [13] Seurat_4.1.1                  </span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##   [1] systemfonts_1.0.4     plyr_1.8.7            igraph_1.3.1         </span>
<span class="co">##   [4] lazyeval_0.2.2        shinydashboard_0.7.2  splines_4.1.3        </span>
<span class="co">##   [7] listenv_0.8.0         scattermore_0.8       ggplot2_3.3.6        </span>
<span class="co">##  [10] digest_0.6.29         htmltools_0.5.2       fansi_1.0.3          </span>
<span class="co">##  [13] magrittr_2.0.3        memoise_2.0.1         googlesheets4_1.0.0  </span>
<span class="co">##  [16] tensor_1.5            cluster_2.1.3         ROCR_1.0-11          </span>
<span class="co">##  [19] globals_0.15.0        matrixStats_0.62.0    pkgdown_2.0.3        </span>
<span class="co">##  [22] spatstat.sparse_2.1-1 colorspace_2.0-3      rappdirs_0.3.3       </span>
<span class="co">##  [25] ggrepel_0.9.1         textshaping_0.3.6     xfun_0.30            </span>
<span class="co">##  [28] dplyr_1.0.9           crayon_1.5.1          jsonlite_1.8.0       </span>
<span class="co">##  [31] progressr_0.10.0      spatstat.data_2.2-0   survival_3.3-1       </span>
<span class="co">##  [34] zoo_1.8-10            glue_1.6.2            polyclip_1.10-0      </span>
<span class="co">##  [37] gargle_1.2.0          gtable_0.3.0          leiden_0.4.2         </span>
<span class="co">##  [40] future.apply_1.9.0    abind_1.4-5           scales_1.2.0         </span>
<span class="co">##  [43] DBI_1.1.2             spatstat.random_2.2-0 miniUI_0.1.1.1       </span>
<span class="co">##  [46] Rcpp_1.0.8.3          viridisLite_0.4.0     xtable_1.8-4         </span>
<span class="co">##  [49] reticulate_1.24       spatstat.core_2.4-2   bit_4.0.4            </span>
<span class="co">##  [52] DT_0.23               htmlwidgets_1.5.4     httr_1.4.3           </span>
<span class="co">##  [55] RColorBrewer_1.1-3    ellipsis_0.3.2        ica_1.0-2            </span>
<span class="co">##  [58] farver_2.1.0          pkgconfig_2.0.3       sass_0.4.1           </span>
<span class="co">##  [61] uwot_0.1.11           deldir_1.0-6          utf8_1.2.2           </span>
<span class="co">##  [64] labeling_0.4.2        tidyselect_1.1.2      rlang_1.0.2          </span>
<span class="co">##  [67] reshape2_1.4.4        later_1.3.0           cellranger_1.1.0     </span>
<span class="co">##  [70] munsell_0.5.0         tools_4.1.3           cachem_1.0.6         </span>
<span class="co">##  [73] cli_3.3.0             generics_0.1.2        ggridges_0.5.3       </span>
<span class="co">##  [76] evaluate_0.15         stringr_1.4.0         fastmap_1.1.0        </span>
<span class="co">##  [79] yaml_2.3.5            ragg_1.2.2            goftest_1.2-3        </span>
<span class="co">##  [82] bit64_4.0.5           knitr_1.38            fs_1.5.2             </span>
<span class="co">##  [85] fitdistrplus_1.1-8    purrr_0.3.4           RANN_2.6.1           </span>
<span class="co">##  [88] pbapply_1.5-0         future_1.25.0         nlme_3.1-157         </span>
<span class="co">##  [91] mime_0.12             formatR_1.12          hdf5r_1.3.5          </span>
<span class="co">##  [94] compiler_4.1.3        rstudioapi_0.13       curl_4.3.2           </span>
<span class="co">##  [97] plotly_4.10.0         png_0.1-7             spatstat.utils_2.3-1 </span>
<span class="co">## [100] tibble_3.1.7          bslib_0.3.1           stringi_1.7.6        </span>
<span class="co">## [103] highr_0.9             desc_1.4.1            rgeos_0.5-9          </span>
<span class="co">## [106] lattice_0.20-45       Matrix_1.4-1          SeuratDisk_0.0.0.9020</span>
<span class="co">## [109] shinyjs_2.1.0         vctrs_0.4.1           pillar_1.7.0         </span>
<span class="co">## [112] lifecycle_1.0.1       BiocManager_1.30.16   spatstat.geom_2.4-0  </span>
<span class="co">## [115] lmtest_0.9-40         jquerylib_0.1.4       RcppAnnoy_0.0.19     </span>
<span class="co">## [118] data.table_1.14.2     cowplot_1.1.1         irlba_2.3.5          </span>
<span class="co">## [121] httpuv_1.6.5          R6_2.5.1              promises_1.2.0.1     </span>
<span class="co">## [124] renv_0.15.4           KernSmooth_2.23-20    gridExtra_2.3        </span>
<span class="co">## [127] parallelly_1.31.1     codetools_0.2-18      MASS_7.3-56          </span>
<span class="co">## [130] assertthat_0.2.1      rprojroot_2.0.3       withr_2.5.0          </span>
<span class="co">## [133] presto_1.0.0          sctransform_0.3.3     mgcv_1.8-40          </span>
<span class="co">## [136] parallel_4.1.3        grid_4.1.3            rpart_4.1.16         </span>
<span class="co">## [139] tidyr_1.2.0           rmarkdown_2.13        googledrive_2.0.0    </span>
<span class="co">## [142] Rtsne_0.16            shiny_1.7.1</span></code></pre>
</details>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Andrew Butler, Charlotte Darby, Yuhan Hao, Paul Hoffman, Rahul Satija, Satija Lab and Collaborators.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
